File: transaction-study/src/main/java/yun/blog/transactionstudy/order/OrderService.java
package yun.blog.transactionstudy.order;

import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;


@Service
@RequiredArgsConstructor
@Transactional
public class OrderService {

  private final OrderRepository orderRepository;
  private final OrderFailedService orderFailedService;

  @Transactional(propagation = Propagation.REQUIRES_NEW)
  public Order create() {

    Order order = save();
    orderFailedService.fail();

    return order;

  }

  public Order save() {
    final Order order = orderRepository.save(new Order(OrderStatus.READY));
//    if (true) {
//      throw new RuntimeException();
//    }
    return order;
  }


  @Transactional(propagation = Propagation.REQUIRES_NEW)
  public Order fail() {
    final Order order = orderRepository.save(new Order(OrderStatus.FAILED));
    if (true) {
      throw new RuntimeException();
    }
    return order;
  }

  public List<Order> findAll() {
    return orderRepository.findAll();
  }

}

File: yun/service/service-order/src/main/kotlin/com/cheese/yun/service/order/OrderService.kt
package com.cheese.yun.service.order

import com.cheese.yun.domain.model.Address
import com.cheese.yun.domain.order.Order
import com.cheese.yun.domain.order.OrderRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.math.BigDecimal


@Service
class OrderService(
    private val orderRepository: OrderRepository
) {
    @Transactional
    fun order(
        address: Address,
        price: BigDecimal
    ) {
        orderRepository.save(Order(address, price))
    }
}